New playground: Visualizing Quantum State

indisoluble opened the pull request:
### What
Partial implementation of the [jupyter notebook with the same name](https://github.com/QISKit/qiskit-tutorial/blob/master/1_introduction/visualizing_quantum_state.ipynb).

**This pull request if not ready to be merged** (qsphere and Bloch sphere are not implemented yet) but I would like a second opinion before going any further.

### Why
In order to replicate this notebook, I needed to re-implement [visualization.py](https://github.com/QISKit/qiskit-sdk-py/blob/master/qiskit/tools/visualization.py) for Swift. This file is based on [matplotlib](https://matplotlib.org/) but, as far as I know, there is no matplotlib equivalent for Swift, so I ended up using this javascript library: [echarts](https://ecomfe.github.io/echarts-doc/public/en/index.html) (more exactly, [echarts 4.0.2](https://github.com/ecomfe/echarts) & [echarts-gl 1.0.2](https://github.com/ecomfe/echarts-gl)).

I would like to confirm it is OK to rely on echarts for plotting graphs and, if so, should I move the code I wrote to a `tools` folder inside the **QISKit framework**?

### How
To present a graph:
1. First, we call one of the functions in [Visualization.swift](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/tutorial/VisualizingQuantumState.playground/Sources/Visualization.swift). Here, the data provided is transformed into a format easier to plot on screen.
2. Then, the new data is passed to an instance of [Barchart.swift](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/tutorial/VisualizingQuantumState.playground/Sources/Barchart.swift) or [Barchart3D.swift](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/tutorial/VisualizingQuantumState.playground/Sources/Barchart3D.swift) which in turn load a [WKWebView](https://developer.apple.com/documentation/webkit/wkwebview).
3. The WKWebView import the echarts javascript library and asynchronously present the graph.

On regard to the playground itself, because the graphs are asynchronously loaded, they have to be presented in the playground live view. Given that there are a few of them to show: histogram, state ‘city’ & state ‘paulivec’, I also prepared a custom live view called [ContainerLiveView](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/tutorial/VisualizingQuantumState.playground/Sources/ContainerLiveView.swift).

So, there are a couple of things to take into account:
* This code only works in iOS & macOS, WebKit & Cocoa are not available in Linux.
* About the licenses, ECharts is available under the BSD 3 license: [echarts 4.0.2](https://github.com/ecomfe/echarts/blob/master/LICENSE) & [echarts-gl 1.0.2](https://github.com/ecomfe/echarts-gl/blob/master/LICENSE). I think this is compatible with QISKit [Apache License 2.0](https://github.com/QISKit/qiskit-sdk-swift/blob/master/LICENSE) but I am not 100% sure.

indisoluble committed:
Plot histogram and state ‘city’, ‘paulivec’, ‘bloch’ & ‘qsphere’ (for macOS/iOS)

manoelmarques commented:
Hi @indisoluble 

As long as the code is all contained in the playground and not part of the main SDK I don't see a problem if it only runs in iOS & macOS. As for the License I am not sure. I am asking internally.
Thank you

indisoluble commented:
Hi @manoelmarques,

Yes, so far all the code is contained in the playground (and its subfolders). But I mentioned that it only runs in iOS & macOS because I was wondering if, regardless of it, I could include the code in the main SDK. However, based on what you just said, it seems that is not possible, doesn't it? :(

manoelmarques commented:
Hi @indisoluble,

So far I have managed to keep the main SDK consistent among platforms. In principle if it gets added to the main SDK, I would need to have some kind of equivalent code for Ubuntu. Anyway, I will ask internally what the future directions are.

Thanks

indisoluble commented:
Thanks for the answer @manoelmarques and for making these questions internally.

I will see what I can do to provide a solution for Ubuntu. Well, first I am going to try to plot a *qsphere* and *bloch sphere* with what I already have and then I will try to port it to Linux :)

Also, this is still a work in progress but do you mind if I keep this PR open? I might need to ask you a few more things.

manoelmarques commented:
No problem,

It will be great if you can help with Linux. Thanks a lot for your help.

indisoluble commented:
Hi @manoelmarques,

I said on my last comment that I would try to plot a *qsphere* and *bloch sphere* before seeing what I could with Linux but, to plot a *qsphere*, I need to [get the eigenvalues and eigenvectors](https://github.com/QISKit/qiskit-sdk-py/blob/master/qiskit/tools/visualization.py#L331) which is not currently available in Swift. So I decided to put this on a side and study how to replicate what I already have on Linux instead.

While on macOS/iOS I use a [WKWebView](https://developer.apple.com/documentation/webkit/wkwebview) to render the graphs, I think I can do the same in Linux using a [WebKitGTK+ WebView](https://webkitgtk.org/reference/webkitgtk/stable/webkitgtk-webkitwebview.html). To probe if it was possible I made this Swift demo app: [CWebkitGtk-Linux-Demo](https://github.com/indisoluble/CWebkitGtk-Linux-Demo), have a look when you a moment.

So, I need you to confirm a couple of things:
1. Is it OK to plot the graphs with a  [WKWebView](https://developer.apple.com/documentation/webkit/wkwebview) in macOS/iOS and a [WebKitGTK+ WebView](https://webkitgtk.org/reference/webkitgtk/stable/webkitgtk-webkitwebview.html) in Linux (Ubuntu)?
2. Is ECharts BSD 3 license: [echarts 4.0.2](https://github.com/ecomfe/echarts/blob/master/LICENSE) & [echarts-gl 1.0.2](https://github.com/ecomfe/echarts-gl/blob/master/LICENSE), compatible with QISKit [Apache License 2.0](https://github.com/QISKit/qiskit-sdk-swift/blob/master/LICENSE)?

If the answer to both question is yes, I will start to work in a new PR that will include the necessary changes in the main SDK. Otherwise, it would be a pity to lose this code so I would move it to a new repo and offer it as a Swift package.

manoelmarques commented:
Hi @indisoluble,

Sorry for the delay, we are all very happy with your contribution but are waiting internal answer on the license issue. We should have an answer on it very soon. 
As for the item 1 above I am completely fine with your solution but I was wondering how will you run it on Ubuntu ? I thought Playgrounds were not implemented there.

indisoluble commented:
Yeah, as far as I know playgrounds are only available on iOS/macOS.

To be honest I did not think much about it but now that you mention it, what about a command-line app for Linux? To develop CWebkitGtk-Linux-Demo I used an editor with a terminal embedded, after making a change I run the code by executing swift run, I found it quite convenient for simple apps.

Anyway if the solution I proposed is fine, I think I will start to prepare the new changes and move the bulk of the code to the main sdk. If the licenses and finally incompatible, I can always copy&paste the code in a new repo (hopefully it will be ok)

> On 20 Feb 2018, at 15:05, Manoel Marques <notifications@github.com> wrote:
> 
> Hi @indisoluble,
> 
> Sorry for the delay, we are all very happy with your contribution but are waiting internal answer on the license issue. We should have an answer on it very soon.
> As for the item 1 above I am completely fine with your solution but I was wondering how will you run it on Ubuntu ? I thought Playgrounds were not implemented there.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

manoelmarques commented:
Hi @indisoluble,

It is fine as command line app in Linux. It would be great if you could have the linux command line app in the tutorial folder too. As for adding to the main SDK, you may have noticed that the Swift SDK is a port of the main Python SDK. I have tried to keep same folder structures, class names, methods etc so that it is more manageable for the many changes I need to make to keep both SDK's in sync. The Visualization in Python is under qiskit/tools folder so it would be great if you could have it there also.
Thanks.

indisoluble commented:
Sure, no problem.

I'll start with it and come back to you ASAP.

manoelmarques commented:
Hi @indisoluble,

Good news. Just received the answer. As long as you include the original LICENSE file where the echart files are, everything should be ok.

indisoluble commented:
Hi @manoelmarques,

I already moved the code to visualise a state to the framework. It is now compatible with iOS as well as macOS but I am still working on the changes for Linux. However, things are going ok and I am confident that along this week it will be ready.

By the way, I noticed there is another project based on javascript: [qiskit-sdk-js](https://github.com/QISKit/qiskit-sdk-js), should we include them in the review? (given that I am using a javascript library to render the graphs).

indisoluble commented:
I am afraid things were going too well, I just found a *major* problem.

I put inside this folder: [resources](https://github.com/indisoluble/qiskit-sdk-swift/tree/master/Sources/qiskit/tools/visualization/resources), the assets used to render the graphs:
* Htmls
* Css
* Javascript libraries (+ licenses)
 
But *Swift Package Manager* can not handle resources: [[SwiftPM] Need a way to include resources with targets](https://bugs.swift.org/browse/SR-2866). Only source files can be included in a Swift package.

I could use this tool: [FakeBundle](https://github.com/zweigraf/FakeBundle), to transform the entire folder in a Swift class but I would rather build the html programatically.

What do you think? @manoelmarques

manoelmarques commented:
Hi @indisoluble 

I had the same problem with some QASM files I wanted to use for unit testing. I ended up putting all of them inside structs with static constants: Tests/qiskitTests/qasm/  That wasn't so bad because I kept them as strings and preserved the whole body. I just grabbed each file added the header bottom part and changed the file type to swift. Inside the code, I just access the constant strings and load them.

indisoluble commented:
Great but I have another question.

I can edit the JavaScript libraries and turn them into Swift files but what should I do with the licenses? I mean, they have to be next to the libraries and for that reason, I copied them in the same folder the libraries are and I also copy them into the QISKit library when it is compiled (I am not sure if the second step is necessary but I preferred to play safe).

If we are turning the libraries into swift files, I think I can simply add at the top of the new files the corresponding license. What do you think @manoelmarques? 

Sorry I did not mention this before, I noticed it after posting my last comment.

> On 26 Feb 2018, at 00:39, Manoel Marques <notifications@github.com> wrote:
> 
> Hi @indisoluble
> 
> I had the same problem with some QASM files I wanted to use for unit testing. I ended up putting all of them inside structs with static constants: Tests/qiskitTests/qasm/ That wasn't so bad because I kept them as strings and preserved the whole body. I just grabbed each file added the header bottom part and changed the file type to swift. Inside the code, I just access the constant strings and load them.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

manoelmarques commented:
Hi @indisoluble 

Actually including the license file inside the folder where the code is was the first thing the lawyers we consulted asked for. :-)
You can put the license file in the library folder and have Swift Package Manager exclude it:
From the Swift Package Manager docs:
"
exclude
Use this property to exclude files and directories from the package sources.
Every item specifies a relative path from the root of the package.
let package = Package(
    name: "Foo",
    exclude: ["Sources/Fixtures", "Sources/readme.md", "Tests/FooTests/images"]
)
This is helpful when you want to place files like resources or fixtures that should not be considered by the convention system as possible sources."

indisoluble commented:
Hi @manoelmarques,

Finally, the code in this PR is compatible with macOS, iOS & Linux. Please, have a look and let me know what you think.

Tips:
- The [demo app for Linux](https://github.com/indisoluble/qiskit-sdk-swift/tree/master/tutorial/VisualizingQuantumState-Linux) relies on SPM to import the QISKit framework and, although [the package is included with a relative path](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/tutorial/VisualizingQuantumState-Linux/Package.swift#L10), it is mandatory to specify a **tag**. What I am trying to say is that, in order to test the app, if you download my own repo you will have to tag the last commit in master branch with `0.1.0`; if you merge this PR in the main repo, you will have to tag the last commit with the former tag afterward, otherwise `swift build` will fail.
- We talked about including the javascript libraries as string constants in the framework but these libraries include escape characters that make impossible to envelop them in `""" ... """`. Instead I transformed the JS files to base64 encoded strings and I decode them as needed.
- If you run the [demo app](https://github.com/indisoluble/qiskit-sdk-swift/tree/master/tutorial/VisualizingQuantumState-Linux) and [the playground](https://github.com/indisoluble/qiskit-sdk-swift/tree/master/tutorial/VisualizingQuantumState.playground) you will notice that the resulting charts are different. Given that the code in both cases is the same, I am assuming the different is due to these warnings printed in the log for the linux app:
```
"Warning have dropped measure from unitary simulator"
"Warning have dropped measure from unitary simulator"
"Warning have dropped measure from unitary simulator"
"Warning have dropped measure from unitary simulator"
"Warning have dropped measure from unitary simulator"
"Warning have dropped measure from unitary simulator"
```

To end, I will include some comments on the files to easy the review.

indisoluble commented on Sources/qiskit/datastructures/Matrix+ColumnRowValueArray.swift:
```diff
@@ -0,0 +1,28 @@
+// =============================================================================
+
+extension Matrix {
+    public func columnRowValueArray() -> [[Any]] {
```
Used [to plot a state with method "city"](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/Sources/qiskit/tools/visualization/Visualization.swift#L65).

indisoluble commented on Sources/qiskit/datastructures/String+Base64.swift:
```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+extension String {
+    func base64Decoded() -> String? {
```
Used to decode the javascript libraries [here](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/Sources/qiskit/tools/visualization/charts/BarChartHtmlFactory.swift#L37) & [here](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/Sources/qiskit/tools/visualization/charts/StateCityChartHtmlFactory.swift#L42).

indisoluble commented on Sources/qiskit/datastructures/String+ZFill.swift:
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+extension String {
+    public func zfill(_ length: Int) -> String {
```
Used to [plot a state with method "city"](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/Sources/qiskit/tools/visualization/Visualization.swift#L62).

indisoluble commented on Sources/qiskit/tools/visualization/charts/platforms/LinuxWebViewFactory.swift:
```diff
@@ -0,0 +1,72 @@
+
+        static func makeWebView(html: String) -> VisualizationTypes.View {
+            let widget = webkit_web_view_new()!
+            enableWebGL(in: widget)
```
3D graphs, like the "city" chart, requires *WebGL*.

indisoluble commented on tutorial/VisualizingQuantumState-Linux/Package.swift:
```diff
@@ -0,0 +1,19 @@
+    name: "VisualizingQuantumState-Linux",
+    dependencies: [
+        // Dependencies declare other packages that this package depends on.
+        .package(url: "../../", from: "0.1.0"),
```
Notice the tag `0.1.0`.

indisoluble commented on tutorial/VisualizingQuantumState.playground/Contents.swift:
```diff
@@ -0,0 +1,168 @@
+/*:
```
This playground is a very simplified version of the [equivalent jupyter notebook](https://github.com/QISKit/qiskit-tutorial/blob/master/1_introduction/visualizing_quantum_state.ipynb), it lacks many of the explanations included in the former. Hope it is enough.

manoelmarques commented:
Hi @indisoluble,

Sorry for the delay but there is an internal approval process. There is dependency link on your project at the main sdk package, for Linux right?
#if os(Linux)
packageDependencies += [
    .package(url: "https://github.com/indisoluble/CWebkitGtk-Linux.git", from: "1.0.0"),
]
#endif
Unfortunately I didn't get approval for this. I don't know if you would ve willing to include this in the main project...
I have tested only under OSX so far. The visualization is very impressive but I had problems with scrolling the Live View. I had to play a lot with it moving from right to bottom, resetting etc to be able to scroll all the charts. Most of the time, it just shows the last one and I cannot scroll up. Do you see the same or maybe I am missing something ?
macOS 10.13.3 (17D102)
Xcode: Version 9.2 (9C40b)

Thank you, the code looks very impressive.

indisoluble commented:
Hi @manoelmarques,

Glad to hear from you :)

`https://github.com/indisoluble/CWebkitGtk-Linux` is under MIT license, I was hoping there would be no problem including it as a dependency. Anyway, to create this wrapper I followed the instructions here: [Require System Libraries](https://github.com/apple/swift-package-manager/blob/master/Documentation/Usage.md#require-system-libraries), in the official documentation for the SPM. What I mean is that I do not think we can include it in the main repo, instead:
- Could you fork mine into QISKit organization?
- Or just copy it and create it from scratch

As I said is under MIT license, so it is perfectly OK.

Now, about the LiveView ... did you notice a black space at each side of the charts? Each chart is fully interactive, you can scroll each one up and down individually and also you can zoom in and out. Now, to scroll through the entire list of charts in macOS, you have to put your cursor over the black space, this is the background view containing the charts and if you scroll once there, you should be able to go to the first or last chart. In Linux, I use a tabbed view, it is a lot easier (and obvious) to navigate through the different charts. I guess I could do a better job here, if it is a MUST DO let me know.

On the other hand, I just started to work on the *Bloch sphere*.

indisoluble commented:
Hi @manoelmarques,

With the last commits I just pushed, we are now able to plot *Bloch spheres* too.

About [ContainerLiveView](https://github.com/indisoluble/qiskit-sdk-swift/blob/master/tutorial/VisualizingQuantumState.playground/Sources/ContainerLiveView.swift), i.e. the view used to present the charts in the playground. This is a very simple view that stack the graphs to present in the live view one on top of the other. Because it is simple, it is also easy to port to iOS or macOS. Anything more complicated than this, it is going to take a while but I have another idea.

With the code in this PR, we can plot: *histograms* and visualise a state with methods *city*, *paulivec* & *bloch*. But we can not plot a *qsphere* until we implement an algorithm to get the eigenvectors and eigenvalues of a matrix. I would rather focus on this than in improving the *ContainerLiveView*; I mean, it might not be very intuitive but it is functional nonetheless.

So, I still have to change this dependency: *https://github.com/indisoluble/CWebkitGtk-Linux*, to another one inside QISKit organisation. But apart from this and if you agree, I would like to consider this PR completed and ready to be merged (unless you think there are other things that have to be fixed or improved).

I'll start to work on the algorithm to get the eigenvectors and eigenvalues ASAP.

manoelmarques commented:
Hi @indisoluble,

I tested it under Ubuntu also and everything looks wonderful :-). I agree with you that technically this pull request is finished, however I had to inform my manager about the WebkitGtk+ library and the need to have an extra repo for Ubuntu. I will need permission on those two items. I will let you know as soon as I have an answer. Thanks  a lot for your contribution.

indisoluble commented:
Awesome :)

By the way, on the next LTS release of Ubuntu: [17.10](https://www.ubuntu.com/desktop/1710), the default desktop will be GNOME/GTK+. Based on this, hopefully to depend on WebKitGTK+ will not be seen as something troublesome.

manoelmarques commented:
That is great @indisoluble . I will pass that info too. That will definitely easy things

indisoluble commented:
Hi @manoelmarques,

I just added this issue to the repo: [Enhancement: Add dependency LAPACK](https://github.com/QISKit/qiskit-sdk-swift/issues/4).

I am also mentioning it here because:

1. If it is accepted, we would be one step closer to be able to compute the eigenvalues & eigenvectors which is a requirement to plot a *qsphere*, the only missing visualisation method.
2. It requires to create another repo for Ubuntu, similar to the one we need for WebKitGTK+.

Would it be possible to discuss this requirement with your manager along with the requirement for WebKitGTK+?

As with WebKitGTK+, I already created:

- The Swift package: [CLapacke-Linux](https://github.com/indisoluble/CLapacke-Linux)
- A demo app: [EigenSwiftDemo](https://github.com/indisoluble/EigenSwiftDemo)

indisoluble commented:
Hi @manoelmarques,

Sorry to bother you but do you have any news? Did you get permission to create the new repos?

manoelmarques commented:
Hi @indisoluble,

My manager is looking into it. I would have approved it already if not for the fact they prefer to have the  WebKitGTK+ and LAPACK repos under the QISKit group and it takes some explaining to the rest of the group the need for those Swift packages there because of Swift Package Manager requirements.

indisoluble commented:
Oh, I see ... that’s a lot of explaining :)

Thanks for the info. Most appreciated.

indisoluble commented on Sources/qiskit/datastructures/Matrix.swift:
```diff
@@ -443,6 +457,74 @@ extension Matrix where T == Complex {
         return m
     }
 
+    public func eigh() throws -> (Vector<Double>, Matrix<Complex>) {
```
As mentioned in https://github.com/QISKit/qiskit-sdk-swift/pull/5, although this method is based on LAPACK function `zheevd`, it is not as efficient as the latter because it iterates up to 6 times through the entire list of values and its memory footprint is bigger.

indisoluble commented:
Hi @manoelmarques,

It took a while but with the code in this PR, we are finally able to plot *histograms* as well as states with method: *city*, *paulivec*, *bloch spheres* & *qspheres* :)

When testing it on Linux, keep in mind what I said in one of the comments above (because I totally forgot it):

> The demo app for Linux relies on SPM to import the QISKit framework and, although the package is included with a relative path, it is mandatory to specify a tag. What I am trying to say is that, in order to test the app, if you download my own repo you will have to tag the last commit in master branch with 0.1.0; if you merge this PR in the main repo, you will have to tag the last commit with the former tag afterward, otherwise swift build will fail.

So, unless you find a bug or something to improve, the only thing remaining to do is to set the dependencies for Linux: WebKitGTK+ & LAPACK, to repos inside QISKit organisation. About this, hope you do not mind if I ask again but do you have news?

Regards.

manoelmarques commented:
Hi @indisoluble,

I talked again to management last Wednesday about creating 2 new repos (WebKitGTK+ & LAPACK) under the QISKit Organization and I am waiting their response.

manoelmarques commented:
Hi @indisoluble 

This is a great solution. I totally agree. If you do it, then I can close the macOS/iOS PR and leave the Linux PR open until the additional 2 repos get created.
Thank you

indisoluble commented:
Hi @manoelmarques,

Instead of creating a new PR, I removed from this one the code for Linux, i.e. this PR adds visualisation capabilities only for macOS/iOS (it still compiles in Linux, of course).

I moved the code for Linux to this commit: https://github.com/indisoluble/qiskit-sdk-swift/commit/91f55accbf5a50248548fd2d60ffda30d3d463b0. I will create a new PR based on it tomorrow ASAP.

Regards.

gaming-hacker commented:
Does this work? With osx 10.12, it can't find the framework, throws up errors, etc.

manoelmarques commented:
The code has been tested and works with:
Apple Swift version 4.1 (swiftlang-902.0.48 clang-902.0.37.1)
Xcode Version 9.3.1 (9E501)
macOS High Sierra: 10.13.4
Carthage 0.29.0

Before opening the QisSwiftPlayground project you should run 'carthage update' in the tutorial folder. For better see, you should put the Live View below the playground view:
![screen shot 2018-05-15 at 12 10 37 am](https://user-images.githubusercontent.com/16138150/40036290-975eaa26-57d4-11e8-95be-27ee9e837b05.png)
